Support explicit nested using of template commonly used

softprofe opened the pull request:
This change is for support explicit nested using of template, eg vector<optional<int>> or map<int,vector<float>> and more complex type, see EPE559 also.

After this change don't need to force user add typedef optional<int> opt_int and then vector<opt_int >, cdt will automatically add abi type for the mid type optional<int >

<!-- PLEASE FILL OUT THE FOLLOWING MARKDOWN TEMPLATE -->
<!-- PR title alone should be sufficient to understand changes. -->

## Change Description
<!-- Describe your changes, their justification, AND their impact. Reference issues or pull requests where possible (use '#XX' or 'GH-XX' where XX is the issue or pull request number). -->

 vector, map,  pair, optional, tuple, variant  and builtin and alias type can mixed use.
      std::tuple<uint64_t, std::optional< float >, std::vector < int > > tup1;
      std::variant<uint64_t, std::optional<std::pair <int, float > >, std::vector < int > > var1;

However, I find a problem that was not caused by my change.
if you have a type (not nested) map<A,B>  then the cdt will be define a structure pair_A_B { "key" , "vaule" }
and if same time you have a type pair<A,B>, then cdt will define another structure will same name pair_A_B but {"first","second"},
so when it try get pair_A_B, it can only has one sturcture in use, the other one will fail.   

Another conflict I found is that,  for example:
typedef struct _mystruct{
      int field0;
      float field1;
      std::vector<std::vector<std::string>> field2;
   }mystruct;

User can use _mystruct in smart contract everywhere, or user can user mystruct everywhere. but user can't use both _mystruct somewhere and mystruct somewhere, there would be some conflict. just like map and pair.

I think in this PR just let it go, one PR can't solve too many things.

## API Changes
- [ ] API Changes
<!-- checked [x] = API changes; unchecked [ ] = no changes, ignore this section -->
<!-- If this PR introduces API changes, please describe the changes here. What will developers need to know before upgrading to this version? -->


## Documentation Additions
- [x] Documentation Additions
<!-- checked [x] = Documentation changes; unchecked [ ] = no changes, ignore this section -->
<!-- Describe what must be added to the documentation after merge. -->

The mechanism we can support explicit nested template is that we automatically add abi typedef for the middle type for user. as we the function in abigen.hpp detects the any explicit nested template is used, it will parse what is/are the middle types and will add typedef for it, so that user don't need do it in cpp file. 

for example: 
When we use std::vector<std::vector<std::vector < int > > > vvvstr;
then the  std::vector< int > will be regards as middle type as abi will add "B_vector_int32_E" for this type, the B means begin, and the E means end, B and E is useful when the middle type is also nested so we can understand how which are nested by the name string.
 middle type                                               new type name for middle type
std::vector< int >                                                       B_vector_int32_E
std::vector<std::vector< int > >                                  B_vector_ B_vector_int32_E_E
std::optional<std::pair<int, float> >                         B_optional_B_pair_int32_float32_E_E"

and the rule is    a new type name begin from B_ and the first template name, and then with a underscore and the template parameter type name(which can be builtin type (int32, float32)or another middle type (B_vector_int32_E), and finally and underscore and E.  These names are generated add_explicit_nested_type(), parsed by translate_explicit_nested_type().

softprofe committed:
This change is for support explicit nested using of template, eg vector<optional<int>> or map<int,vector<float>> and more complex type, see EPE559 also.
After this change don't need to force user add typedef optional<int> opt_int and then vector<opt_int>, cdt will automatically add abi type for the mid type optional<int>

softprofe committed:
Fix, don't support nested with "ignore" so far, so kick it out.

softprofe committed:
Fix, don't support "index" nested using. kick it out for now.

plroblox commented on tests/unit/test_contracts/explicit_nested_tests.cpp:
```diff
@@ -0,0 +1,106 @@
+   [[eosio::action]]
+    // usage : cleos -v push action eosio var '[["uint64",8]]'  -p eosio@active
+    // usage : cleos -v push action eosio var '[["B_optional_B_pair_int32_float32_E_E", {"first":1,"second":2.0}]]'  -p eosio@active
+    // usage : cleos -v push action eosio var '[["B_vector_int32_E", [2,3,4]]]'  -p eosio@active
```
excellent revelation on std::variant<Ts...> JSON input format !

plroblox commented on tools/include/eosio/abigen.hpp:
```diff
@@ -386,11 +386,224 @@ namespace eosio { namespace cdt {
          evaluated.insert(t.getTypePtr());
          auto type = get_ignored_type(t);
+         if(is_explicit_nested(t)){
+            add_explicit_nested_type(t.getNonReferenceType());
```
What about the outermost layer is not explcitly_nested, but inner layer is explcitly `nested?` e.g
   ```
      typedef vector<vector<vector<int> > > myvec3;
      typedef set<myvec3> myset;
```
Can you handle this scenario ?

softprofe replied:
Yes, I add test contract to test this case, it works well as expected. 
The reason is that in function add_explicit_nested_type judge the case of is_aliasing and have dealing with this case.
This concern is very good of, by add two new test actions we can make sure it works.

And also Qing's concern is very good also, he want me add test action for set, deque and list nested, and I did add this action and it works well also.

plroblox reviewed (commented):
See my 2nd comment

softprofe committed:
Add test contract to test set, deque, list and whole type is not nested but inside type is explicit nested. passed.

plroblox commented on tests/toolchain/abigen-pass/nested_container.abi:
```diff
@@ -1,7 +1,16 @@
-    "types": [],
+    "types": [
+        {
+            "new_type_name": "B_map_string_string_E",
```
I think you need add something in 'Documentation Additions', because new type names are added because of your code change, for example, you can explain the syntax/rule of new-type-name, like B/E represents Beginning/End of a parsing unit, what's going to happen for a mixture of explcitly nested type and typedefined type?  etc.

Also I think this is a significant change pn abi, so we probably need consult with Bucky & Santhosh to update abi/1.2 to abi/1.3 in an additional JIRA ticket

softprofe replied:
Yes, good idea, I added Santhosh to reviewer.

plroblox commented on tests/unit/test_contracts/explicit_nested_tests.cpp:
```diff
@@ -0,0 +1,160 @@
+   typedef set<myvec3> myset;
+
+   [[eosio::action]]   // this test is for the whole type is not explicit nested by the inside type is a alias of explicit nested.
+   // usage : cleos -v push action eosio sivvvi '[[[[[1, 2],[3,4]]]]]' -p eosio@active
```
Awesome! BTW cleos push action 'writes' contract data to nodeos, cleos get table 'reads' data from nodeos, have you verified **cleos get table** also works ?

softprofe replied:
Good concern, I will test it and let you know the result.

softprofe replied:
It works very well for read table, I run command as : cleos -v get table eosio eosio testdata
as then screen print below data, it is exact the data I put into the table.

{
  "rows": [{
      "id": "6138663579460959028",
      "data": [{
          "key": 12,
          "value": 34
        }
      ],
      "data_vec": [{
          "first": 12,
          "second": "1.20000004768371582"
        },{
          "first": 34,
          "second": "3.40000009536743164"
        }
      ],
      "data_vec2": [{
          "first": 56,
          "second": "5.59999990463256836"
        },{
          "first": 78,
          "second": "7.80000019073486328"
        }
      ],
      "data_vec3": [
        "1.10000002384185791",
        "2.20000004768371582",
        "3.29999995231628418"
      ],
      "tup1": {
        "field_0": 12,
        "field_1": "3.40000009536743164",
        "field_2": [
          5,
          6,
          7
        ]
      },
      "var1": [
        "B_vector_int32_E",[
          8,
          9,
          10
        ]
      ]
    },{
      "id": "6138663579460959041",
      "data": [{
          "key": 12,
          "value": 34
        }
      ],
      "data_vec": [{
          "first": 12,
          "second": "1.20000004768371582"
        },{
          "first": 34,
          "second": "3.40000009536743164"
        }
      ],
      "data_vec2": [{
          "first": 56,
          "second": "5.59999990463256836"
        },{
          "first": 78,
          "second": "7.80000019073486328"
        }
      ],
      "data_vec3": [
        "1.10000002384185791",
        "2.20000004768371582",
        "3.29999995231628418"
      ],
      "tup1": {
        "field_0": 12,
        "field_1": "3.40000009536743164",
        "field_2": [
          5,
          6,
          7
        ]
      },
      "var1": [
        "B_vector_int32_E",[
          8,
          9,
          10
        ]
      ]
    }
  ],
  "more": false,
  "next_key": "",
  "next_key_bytes": ""
}

softprofe committed:
Add test (put/get) for nested template used in table, and also divide the two long functions to small functions for good looking.

plroblox commented on tests/unit/test_contracts/explicit_nested_tests.cpp:
```diff
@@ -0,0 +1,219 @@
+
+   [[eosio::action]]   // this test is for the put simple data to table, so as to check the the data read from table is the same
+   // usage : cleos -v push action eosio putd '[eosio]' -p eosio@active
+   //             cleos -v get table eosio eosio testdata
```
looks good

bogniq commented on tools/include/eosio/abigen.hpp:
```diff
@@ -386,11 +386,244 @@ namespace eosio { namespace cdt {
+               add_typedef(inside_type[i]);
+               inside_type_name[i] = get_base_type_name( inside_type[i] );
+            }
+         }
```
This `for` loop is also used in below `add_explicit_nested_pair`, and similarly used in e.g. `add_explicit_nested_tuple` and `add_explicit_nested_array`(with different `inside_type` sizes). Could it be transformed into a helper function?

softprofe replied:
It's worth to try.

softprofe replied:
Now using adding_explicit_nested_dispatcher instead of the if else group everywhere.

softprofe committed:
Support nested template used with structure like vector<vector<mystruct>>, inside mystruct has nested template is supported also.
And also factor out the "if else cohort" to a helper function so as to reuse them.

plroblox commented on tools/include/eosio/abigen.hpp:
```diff
@@ -399,8 +396,25 @@ namespace eosio { namespace cdt {
          } else if (is_aliasing(inside_type)) { // inside type is an alias
             add_typedef(inside_type);
             inside_type_name = get_base_type_name( inside_type );
+         }   else if (is_template_specialization(inside_type, {})) {
```
yeah, I was also thinking using something like is_struct() right after is_aliasing to handle `vector<vector<mystruct> >`, where is_struct() covers template struct and non-template struct, your implementation here is good.

Strictly speaking, template struct like `vector<vector<mystruct2<string> > >` where `mystruct2<T> i`s a template struct, should also be verified, however this would be come way more complicated, let another JIRA ticker take care of that case

softprofe replied:
Agree with do it (vector<vector<mystruct2<string> > ) in another PR, this PR is already complex.

softprofe replied:
And things like is_struct() , together with different kind of structure like mystruct2<T> or more complex struct type ,  the structure itself is worth another PR, I think you may file a EPE to note this.
